Makefile

```diff
@@ -704,32 +694,81 @@ DEFAULT_FLAGS = \
   -bundle Lib.RandomBuffer.System \
-  -bundle '\*[rename=Misc]' \
-  -minimal -wasm
-
-# Customizations for WASM.
-# - only keep definitions reachable from Test.NoHeap -- this indicates what we
-#   should retain for the WASM distribution.
-dist/wasm/Makefile.basic: TEST_FLAGS = -d wasm
+  -bundle Lib.Memzero \
+  -minimal -wasm -d wasm
+
+dist/wasm/Makefile.basic: VALE_ASMS =
+dist/wasm/Makefile.basic: HAND_WRITTEN_OPTIONAL_FILES =
+dist/wasm/Makefile.basic: HAND_WRITTEN_H_FILES =
+dist/wasm/Makefile.basic: HACL_OLD_FILES =
+dist/wasm/Makefile.basic: HAND_WRITTEN_FILES =
+dist/wasm/Makefile.basic: TARGETCONFIG_FLAGS =
+dist/wasm/Makefile.basic: INTRINSIC_FLAGS =
+
+# Must appear early on because of the left-to-right semantics of -bundle flags.
 dist/wasm/Makefile.basic: HAND_WRITTEN_LIB_FLAGS = $(WASM_FLAGS)
-dist/wasm/Makefile.basic: BUNDLE_FLAGS =
-dist/wasm/Makefile.basic: WASMSUPPORT_BUNDLE =
 
+# Overriding EverCrypt.Hash so that is it no longer a live root; it will be
+# eliminated via the -bundle EverCrypt.* below
+dist/wasm/Makefile.basic: E_HASH_BUNDLE =
+
+# Doesn't work in WASM because one function has some heap allocation
+dist/wasm/Makefile.basic: HASH_BUNDLE += -bundle Hacl.HMAC_DRBG
+
+# Doesn't work in WASM because un-materialized externals for AES128
+dist/wasm/Makefile.basic: FRODO_BUNDLE = -bundle Hacl.Frodo.KEM,Frodo.Params,Hacl.Impl.Frodo.*,Hacl.Impl.Matrix,Hacl.Frodo.*,Hacl.Keccak,Hacl.AES128
+
+# Doesn't work in Wasm because it uses assembler intrinsics
+ECDSA_BUNDLE = -bundle Hacl.Impl.ECDSA,Hacl.Impl.ECDSA,Hacl.Impl.ECDSA.*,Hacl.Impl.P256.*,Hacl.Impl.P256,Hacl.Spec.P256.*,Hacl.Impl.SolinasReduction,Hacl.Impl.LowLevel
+
+# No Vale Curve64; no "Local" Curve64 (local Makefile hack); only the slow 64
+# one
+dist/wasm/Makefile.basic: CURVE_BUNDLE = $(CURVE_BUNDLE_BASE) \
+  -bundle Hacl.Curve25519_64 \
+  -bundle Hacl.Curve25519_64_Local
+
+# Disabling vectorized stuff
+dist/wasm/Makefile.basic: CHACHA20_BUNDLE += \
+  -bundle Hacl.Chacha20_Vec128,Hacl.Chacha20_Vec256
+dist/wasm/Makefile.basic: CHACHAPOLY_BUNDLE += \
+  -bundle Hacl.Chacha20Poly1305_128,Hacl.Chacha20Poly1305_256
+dist/wasm/Makefile.basic: POLY_BUNDLE = \
+  -bundle 'Hacl.Poly1305_32=Hacl.Impl.Poly1305.Field32xN_32' \
+  -bundle 'Hacl.Poly1305_128,Hacl.Poly1305_256,Hacl.Impl.Poly1305.*'
+
+# And Merkle trees
+dist/wasm/Makefile.basic: MERKLE_BUNDLE = -bundle 'MerkleTree,MerkleTree.*'
+dist/wasm/Makefile.basic: CTR_BUNDLE =
+dist/wasm/Makefile.basic: DEFAULT_FLAGS += -bundle 'EverCrypt,EverCrypt.*'
+
+dist/wasm/package.json: dist/wasm/Makefile.basic $(wildcard bindings/js/*.js) bindings/js/README.md $(wildcard bindings/js/*.json) bindings/js/.npmignore
+	cp -f $(filter-out %.basic,$^) $(dir $@)
+	rm $(dir $@)/README
+
+dist/wasm/doc/readable_api.js: dist/wasm/package.json
+	cd dist/wasm && \
+	mkdir -p doc && \
+	node api_doc.js
+
+dist/wasm/doc/out/index.html: dist/wasm/doc/readable_api.js
+	jsdoc $< -d $(dir $@)
+
+wasm-doc: dist/wasm/doc/out/index.html
```

Comment:
can the CI target depend on wasm-doc so that we can upload the JS documentation on github, too?

Reply:
I just did that but then `jsdoc` is not installed (hence the failing build), so I can't generate doc with CI.